repos/system_upgrade/common/libraries/fetch.py

```diff
@@ -108,3 +121,59 @@ def read_or_fetch(filename, directory="/etc/leapp/files", service=None, allow_em
         sp=service_path, l=len(response.content)))
 
     return response.content.decode(encoding)
+
+
+def load_data_asset(actor_requesting_asset,
+                    asset_filename,
+                    asset_fulltext_name,
+                    docs_url,
+                    docs_title):
+    """
+    Load the contents of the data asset with given asset_filename
+
+    :param Actor actor_requesting_asset: The actor instance requesting the asset file. It is necessary for the actor
+                                         to be able to produce ConsumedDataAsset message in order for leapp to be able
+                                         to uniformly report assets with incorrect versions.
+    :param str asset_filename: The file name of the asset to load.
+    :param str asset_fulltext_name: A human readable asset name to display in error messages.
+    :param str docs_url: Docs url to provide if an asset is malformed or outdated.
+    :param str docs_title: Title of the documentation to where `docs_url` points to.
+    :returns: A dict with asset contents (a parsed JSON), or None if the asset was outdated.
+    """
+
+    # Check that the actor that is attempting to obtain the asset meets the contract to call this function
+    # @Note(mhecko): Assertion is fine, as the assertion enforces a 'static' contract between the callee and the caller
+    assert models.LeappRepositoryInfo in actor_requesting_asset.consumes
```

Comment:
@pirat89 what is our position on using asserts outside of tests?

Reply:
It is a general no, no

Asserts should not be used, stopactorexecutionerror is your friend here

Reply:
It is not even needed from the library to _enforce_ that the actors using the library are consuming specific content, so the assertion can be dropped all together, especially when I consider that there will not be a lot of people writing custom actors which use data files. On the other hand, if the code _assumes_ something it should be explicit, and thus I would rather make StopActorExecution there.

My curiosity then stumbles upon the reasons against using assertions. Why shouldn't one use assertions in this particular scenario?

Reply:
long story short:
when you want to use something what is reliable, do not use assert to stop the execution in any code; it'S good for experimenting, testing, ... not for the use in something what you expect to be reliable or have a defined behaviour on all machines.

short story long:
asserts should be really used just in tests. Using asserts in the application itself is bad practice - usually inherited from old programming styles / projects. When you specify assert, you even do not know whether it will be really executed, as you can run python (or whatever language) in the way that all asserts are stripped or ignored. This leads to the situations when the code execution could lead to undefined situations / various different behaviours, and it could be then funny to debug something on your system why is something happening, when someone years ago thought that the problem is clearly caught by an assert. It's also nice source of security issues :)

Some examples:
* https://python.plainenglish.io/if-you-are-using-assert-you-must-read-this-7256e46dc816
* https://towardsdatascience.com/2-caveats-to-avoid-when-using-python-assert-statement-8060e123c26a

Reply:
Thanks for a nice explanation, TIL you can disable assertions globally.

I do not think that the purpose of the asserts in the snippet might lead to the problems described in the linked articles as they were not meant to validate user supplied data (or any kind of data for that matter). Their purpose was to provide a helping hand to the programmer trying to rely on the library, by making sure that the actor in which the library is used consumes and produces everything necessary. The assertions are not even needed -- if an actor does not consume what is needed, then the library will raise a StopActorException about not receiving any LeappRepositoryRPMInfo. Thus the programmer has to think why was the exception raised whereas now the library tells him the first time he tries executing the actor containing a call to that function that he might be  forgetting to specify something in actor's consumes/produces. Leapp end users should not achieve anything by disabling assertions, because it is there to help the developer rather than to make sure that some data are not as expected.

That being said, using whatever kind of exception is fine as long it provides the same value to the developer. I would be even fine with dropping the exception all together based on a sole argument of how many custom actors might want to load data assets. The assertions have been removed in favor of StopActorExecution with the latest push.